### DISCUSSION TITLE
Dynamic port binding for localhost

### DISCUSSION THREAD

**Discussion by jackj93** [dotnet/aspnetcore#26180] [General]
I have the requirements of dynamically binding the same open port to both the IPv4 and IPv6 loopback address.
Our solution currently is demonstrated in the following [blog](https://web.archive.org/web/20130909044601/http://blogs.msdn.com/b/malarch/archive/2005/11/18/494769.aspx).

The problem with this is that once it finds the open port, we have to close the connection/socket before passing the port number to Kestrel.
I know that currently binding to localhost will port bind both the IPv4 and IPv6 addresses but is there a reason why dynamic port binding is not available here?

**Comment by Tratcher** [comment]
Kestrel uses the same trick internally to bind to port 0 when you give it just an IP, or anything except localhost. That doesn't work for localhost because it's actually two bindings for `127.0.0.1` and `::1`. If we tried to bind to port 0 for both we'd be unlikely to get the same port. We also can't use IPv6 dual mode sockets to listen for both IPv4 and IPv6 because that doesn't apply to loopback, they're two unique addresses.

**Reply by jackj93** [reply]
So what happens when I try the above method to get an open (non-IPv6 only) IPv6 port.
Then close the socket and passing the port number to be bound using Kestrel.

Will that bind both addresses to the same port? (I'm assuming it does because of the following [announcement](https://github.com/aspnet/Announcements/issues/185))

If it does, is there a way within the framework (ASP.NET Core) for it to automatically do that?

**Reply by Tratcher** [reply]
Yes, it will bind to both addresses with the same port, but there's no guarantee that the IPv4 port will be available, you only checked the IPv6 port. For that matter, there's no guarantee the IPv6 port will still be available, there's a race condition where it could be taken between the time you close the socket and start kestrel. That's why we don't do it in the framework either.
